### Data Loading and Inspection

In [381]:
# Load dataset:
df_support <- read.csv("support2.csv", header = TRUE, sep = ",")

df_description <- read.csv("support-variables-description.csv", header = TRUE, sep = ",")


In [382]:
# df_description


In [383]:
# Inspect dataset
head(df_support, 3)


age      death sex    hospdead slos d.time dzgroup     dzclass           
1 62.84998 0     male   0         5   2029   Lung Cancer Cancer            
2 60.33899 1     female 1         4      4   Cirrhosis   COPD/CHF/Cirrhosis
3 52.74698 1     female 0        17     47   Cirrhosis   COPD/CHF/Cirrhosis
  num.co edu ... crea     sod ph       glucose bun urine adlp adls
1 0      11  ... 1.199951 141 7.459961 NA      NA  NA     7   7   
2 2      12  ... 5.500000 132 7.250000 NA      NA  NA    NA   1   
3 2      12  ... 2.000000 134 7.459961 NA      NA  NA     1   0   
  sfdm2            adlsc
1                  7    
2 <2 mo. follow-up 1    
3 <2 mo. follow-up 0

In [384]:
# Check structure of dataset
str(df_support)
dim(df_support)


'data.frame':	9105 obs. of  47 variables:
 $ age     : num  62.8 60.3 52.7 42.4 79.9 ...
 $ death   : int  0 1 1 1 0 1 1 1 1 1 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: int  0 1 0 0 0 1 0 0 0 0 ...
 $ slos    : int  5 4 17 3 16 4 9 7 12 8 ...
 $ d.time  : int  2029 4 47 133 2029 4 659 142 63 370 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ num.co  : int  0 2 2 2 1 1 1 3 2 0 ...
 $ edu     : int  11 12 12 11 NA 14 14 NA 12 11 ...
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 2 5 5 1 1 3 1 1 3 ...
 $ scoma   : int  0 44 0 0 26 55 0 26 26 0 ...
 $ charges : num  9715 34496 41094 3075 50127 ...
 $ totcst  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ totmcst : num  NA NA NA NA NA NA NA NA NA NA ...
 $ avtisst : num  7 29 13 7 18.7 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ sps

[1] 9105   47

After Inspecting closely, We see that for many categorical variables', the datatype are not 'Factor' eg. $ sex : chr  "male" "female" etc.

### Data Cleaning and Preprocessing

In [385]:
# Identify and Group the categorical variables as categories of interest
cat_interest <- c("death", "sex", "hospdead", "dzgroup", "dzclass", "income", "race", "diabetes", "dementia", "ca", "dnr", "adlp", "adls", "sfdm2")

# Apply unique function to check unique categories for each variable and check their structure
unique_values <- sapply(df_support[cat_interest], unique)
unique_values

str(unique_values)


$death
[1] 0 1

$sex
[1] male   female
Levels: female male

$hospdead
[1] 0 1

$dzgroup
[1] Lung Cancer       Cirrhosis         ARF/MOSF w/Sepsis Coma             
[5] CHF               Colon Cancer      COPD              MOSF w/Malig     
8 Levels: ARF/MOSF w/Sepsis CHF Cirrhosis Colon Cancer Coma ... MOSF w/Malig

$dzclass
[1] Cancer             COPD/CHF/Cirrhosis ARF/MOSF           Coma              
Levels: ARF/MOSF Cancer Coma COPD/CHF/Cirrhosis

$income
[1] $11-$25k   under $11k            $25-$50k   >$50k     
Levels:  $11-$25k $25-$50k >$50k under $11k

$race
[1] other    white    black    hispanic asian            
Levels:  asian black hispanic other white

$diabetes
[1] 0 1

$dementia
[1] 0 1

$ca
[1] metastatic no         yes       
Levels: metastatic no yes

$dnr
[1] no dnr                          dnr after sadm  dnr before sadm
Levels:  dnr after sadm dnr before sadm no dnr

$adlp
[1]  7 NA  1  0  2  3  5  6  4

$adls
[1]  7  1  0  2 NA  5  4  6  3

$sfdm2
[1]                     <2 mo. follow-up    no(M2 and SIP pres)
[4] SIP>=30             adl>=4 (>=5 if sur) Coma or Intub      
6 Levels:  <2 mo. follow-up adl>=4 (>=5 if sur) ... SIP>=30

List of 14
 $ death   : int [1:2] 0 1
 $ sex     : Factor w/ 2 levels "female","male": 2 1
 $ hospdead: int [1:2] 0 1
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 1 5 2 4 6 8
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 1 3
 $ income  : Factor w/ 5 levels "","$11-$25k",..: 2 5 1 3 4
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 3 4 2 1
 $ diabetes: int [1:2] 0 1
 $ dementia: int [1:2] 0 1
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 3
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 2 3
 $ adlp    : int [1:9] 7 NA 1 0 2 3 5 6 4
 $ adls    : int [1:9] 7 1 0 2 NA 5 4 6 3
 $ sfdm2   : Factor w/ 6 levels "","<2 mo. follow-up",..: 1 2 5 6 3 4


In [386]:
# convert categorical and ordinal variables to factors  # nolint
df_support$sex <- factor(df_support$sex)

df_support$death <- factor(df_support$death)

df_support$hospdead <- factor(df_support$hospdead)

df_support$dzgroup <- factor(df_support$dzgroup)

df_support$dzclass <- factor(df_support$dzclass)

df_support$income <- factor(df_support$income, ordered = TRUE, levels = c("under $11k", "$11-$25k", "$25-$50k", ">$50k"))

df_support$race <- factor(df_support$race)

df_support$diabetes <- factor(df_support$diabetes)

df_support$dementia <- factor(df_support$dementia)

factor(c("metastatic", "no", "yes"),
    levels = c("no", "yes", "metastatic"),
    ordered = TRUE
)

df_support$adlp <- factor(df_support$adlp, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$adls <- factor(df_support$adls, ordered = TRUE, levels = c(1, 2, 3, 4, 5, 6, 7))

df_support$sfdm2 <- factor(df_support$sfdm2, ordered = TRUE, levels = c("no(M2 and SIP pres)", "adl>=4 (>=5 if sur)", "SIP>=30", "Coma or Intub", "<2 mo. follow-up"))


[1] metastatic no         yes       
Levels: no < yes < metastatic

In [387]:
# Check structure of dataset to confirm changes
str(df_support[cat_interest])


'data.frame':	9105 obs. of  14 variables:
 $ death   : Factor w/ 2 levels "0","1": 1 2 2 2 1 2 2 2 2 2 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: Factor w/ 2 levels "0","1": 1 2 1 1 1 2 1 1 1 1 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ income  : Ord.factor w/ 4 levels "under $11k"<"$11-$25k"<..: 2 2 1 1 NA NA 3 NA NA 3 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 4 ...
 $ diabetes: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 2 1 1 ...
 $ dementia: Factor w/ 2 levels "0","1": 1 1 1 1 1 1 1 1 2 1 ...
 $ ca      : Factor w/ 3 levels "metastatic","no",..: 1 2 2 1 2 2 2 2 1 1 ...
 $ dnr     : Factor w/ 4 levels "","dnr after sadm",..: 4 1 4 4 4 4 4 4 2 4 ...
 $ adlp    : Ord.factor w/ 7 levels "1"<"2"<"3"<"4"<..: 7 NA 1 NA NA NA NA NA NA NA ...
 $ adls    : Ord.factor w/ 7 levels "1"<"2"<"3"<

### MISSING VALUES ANALYSIS



In [388]:
# Count the total number of complete cases (Rows with no missing values)
num_complete_cases <- sum(complete.cases(df_support))
num_complete_cases

dim(df_support)

# We have only 106 complete cases (rows with no missing values) out of 9105 rows. This is a very small number of complete cases. In order to make the dataset usable, we will have to find identify which variables have the most missing values and drop them from the dataset.


[1] 106

[1] 9105   47

In [389]:
# Check for Missing Values in each column and sort in descending order

na_counts_col <- colSums(is.na(df_support))

na_counts_col <- sort(na_counts_col, decreasing = TRUE)

na_counts_col[1:15]


adlp    adls   urine glucose     bun totmcst     alb  income    bili    pafi 
   7490    5975    4862    4500    4352    3475    3372    2982    2601    2325 
     ph   prg2m     edu   prg6m   sfdm2 
   2284    1649    1634    1633    1400

Variables with the most missing values are:
1. adlp : 7490 (Too many missing values, and highly correlated with other 'Activities of Daily Living Index' variable (adlsc) 
2. adls : 7490 (same as above)
3. urine : 4862
4. glucose : 4500
5. bun : 4352
6. totmcst (Total micro cost): 3475 (also, Highly correlated with totcst: total cost)


To start, We will drop the variables with more than 50% missing values. These are:
1. adlp, adls, urine, totmcst, glucose, bun.



In [390]:
# Drop variables with high missing values from the list

# And 'column_to_drop' is the name of the column you want to drop
df_updated_mis <- subset(df_support, select = -c(adlp, adls, urine, totmcst, glucose, bun))

# Check structure of dataset to confirm changes
str(df_updated_mis)


'data.frame':	9105 obs. of  41 variables:
 $ age     : num  62.8 60.3 52.7 42.4 79.9 ...
 $ death   : Factor w/ 2 levels "0","1": 1 2 2 2 1 2 2 2 2 2 ...
 $ sex     : Factor w/ 2 levels "female","male": 2 1 1 1 1 2 2 2 2 1 ...
 $ hospdead: Factor w/ 2 levels "0","1": 1 2 1 1 1 2 1 1 1 1 ...
 $ slos    : int  5 4 17 3 16 4 9 7 12 8 ...
 $ d.time  : int  2029 4 47 133 2029 4 659 142 63 370 ...
 $ dzgroup : Factor w/ 8 levels "ARF/MOSF w/Sepsis",..: 7 3 3 7 1 5 2 2 7 4 ...
 $ dzclass : Factor w/ 4 levels "ARF/MOSF","Cancer",..: 2 4 4 2 1 3 4 4 2 2 ...
 $ num.co  : int  0 2 2 2 1 1 1 3 2 0 ...
 $ edu     : int  11 12 12 11 NA 14 14 NA 12 11 ...
 $ income  : Ord.factor w/ 4 levels "under $11k"<"$11-$25k"<..: 2 2 1 1 NA NA 3 NA NA 3 ...
 $ scoma   : int  0 44 0 0 26 55 0 26 26 0 ...
 $ charges : num  9715 34496 41094 3075 50127 ...
 $ totcst  : num  NA NA NA NA NA NA NA NA NA NA ...
 $ avtisst : num  7 29 13 7 18.7 ...
 $ race    : Factor w/ 6 levels "","asian","black",..: 5 6 6 6 6 6 6 6 3 

In [391]:
# Recheck the total number of complete cases (Rows with no missing values) after dropping variables with high missing values
num_complete_cases <- sum(complete.cases(df_updated_mis))
num_complete_cases


[1] 1838

Still, we have only 1838 complete cases. 

To determine further which variables to drop, we will create a correlation matrix and identify the variables with high correlation.

In [392]:
# Create df with only numeric columns for creating a correlation matrix

numeric_columns <- sapply(df_support, is.numeric)
numeric_df <- df_support[numeric_columns]

# Create a corrplot to visualize the correlation between the variables
# install.packages("corrplot")
library(corrplot)
library(reshape2)

# Assuming df is your data frame
correlation_matrix <- cor(numeric_df, use = "complete.obs")

# Flatten and filter
corr_melted <- melt(correlation_matrix)

# Remove self-correlations and duplicates
corr_melted <- subset(corr_melted, Var1 != Var2)
corr_melted <- corr_melted[!duplicated(t(apply(corr_melted[, 1:2], 1, sort))), ]

# Sort by absolute correlation value
corr_melted <- corr_melted[order(-abs(corr_melted$value)), ]
top_correlations <- head(corr_melted, 200) # Top 20 correlations

print(top_correlations)


        Var1    Var2       value
240  totmcst  totcst  0.96893250
410   surv6m  surv2m  0.96039229
51    dnrday    slos  0.89520435
512    prg6m   prg2m  0.89486469
41    totcst    slos  0.83330115
206   totcst charges  0.83288576
42   totmcst    slos  0.81737481
207  totmcst charges  0.80801559
249   dnrday  totcst  0.78364185
282   dnrday totmcst  0.77069697
342      aps     sps  0.77016228
343   surv2m     sps -0.74601563
889      bun    crea  0.70237278
344   surv6m     sps -0.66952642
40   charges    slos  0.61881780
376   surv2m     aps -0.59697262
309      aps avtisst  0.59393388
216   dnrday charges  0.59262860
274  avtisst totmcst  0.56246251
213     hday charges  0.56230912
412    prg2m  surv2m  0.55476203
178   surv2m   scoma -0.54824604
241  avtisst  totcst  0.54771005
446    prg6m  surv6m  0.53961848
308      sps avtisst  0.52650823
445    prg2m  surv6m  0.52642339
413    prg6m  surv2m  0.51205199
377   surv6m     aps -0.49266915
208  avtisst charges  0.49078406
83     prg

We find that the following variables are highly correlated with other variables:
1. sur6m : surv2m (model predicted 6 months and 2 months survival estimates for patient ) - (Cor = 0.96)
2. totcst : charges (cost to charge ratio : total charges) - (Cor = 0.83)
3. aps (APACHE III day 3 physiology score : SUPPORT physiology score) : sps (predicted SUPPORT physiology score on day 3 )  - (Cor = 0.77)
4. aps : avtisst (Average TISS score - quantifies type and number of intensive care treatments. ) - (Cor = 0.59)
5. sfdm2 (Level of functional disability of the patient in a 1-5 scale. (Naturally correlated with other physiological scores ))

Identify out which other variables are also Survival estimates and check their respective number of missing values.
1. prg2m : Physician’s 2-month survival estimate for patient.
2. prg6m : Physician’s 6-month survival estimate for patient.
3. prg2m : surv2m - (Cor = 0.55)
4. prg6m : surv6m - (Cor = 0.53)


In [393]:
# Missing values for prg2m, prg6m, surv2m, surv6m, totcst, charges, aps, sps, avtisst, sfdm2

# Check for Missing Values in each column and sort in descending order

na_counts_cor <- colSums(is.na(df_support[c("prg2m", "prg6m", "surv2m", "surv6m", "totcst", "charges", "aps", "sps", "avtisst", "sfdm2")]))

na_counts_cor <- sort(na_counts_cor, decreasing = TRUE)

na_counts_cor


prg2m   prg6m   sfdm2  totcst charges avtisst  surv2m  surv6m     aps     sps 
   1649    1633    1400     888     172      82       1       1       1       1

In [394]:
# We drop the 'redundant' and with 'high missing values' variables ie. prg2m, prg6m and surv6m, totcst, sps, avtisst

# And 'column_to_drop' is the name of the column you want to drop
df_updated_mis <- subset(df_updated_mis, select = -c(prg2m, prg6m, surv6m, totcst, sps, avtisst, sfdm2))

# Check complete cases
num_complete_cases <- sum(complete.cases(df_updated_mis))
num_complete_cases


[1] 2517

In [395]:
# Check for Missing Values in df_updated_mis and sort in descending order

na_counts_col <- colSums(is.na(df_updated_mis))

na_counts_col <- sort(na_counts_col, decreasing = TRUE)

na_counts_col[1:15]


alb  income    bili    pafi      ph     edu    wblc charges    crea  dnrday 
   3372    2982    2601    2325    2284    1634     212     172      67      30 
  scoma     aps  surv2m  meanbp     hrt 
      1       1       1       1       1

### Imputing Missing Values

According to the HBiostat Repository (https://hbiostat.org/data/repo/supportdesc, Professor Frank Harrell) the following default values have been found to be useful in imputing missing baseline physiologic data:
Baseline Variable	Normal Fill-in Value
- Serum albumin (alb)	3.5
- PaO2/FiO2 ratio (pafi) 	333.3
- Bilirubin (bili)	1.01
- Creatinine (crea)	1.01
- bun	6.51
- White blood count (wblc)	9 (thousands)
- Urine output (urine)	2502

We will drop the rows with missing values for the following variables:
1. wblc
2. charges
3. crea
4. dnrday
5. scoma
6. aps
7. surv2m
8. meanbp
9. hrt
